qa/os/src/test/java/org/elasticsearch/packaging/test/KeystoreManagementTests.java

```diff
@@ -253,6 +258,81 @@ public void test51WrongKeystorePasswordFromFile() throws Exception {
         }
     }
 
+    /**
+     * Check that we can mount a password-protected keystore to a docker image
+     * and provide a password via an environment variable.
+     */
+    public void test60DockerEnvironmentVariablePassword() throws Exception {
+        assumeTrue(distribution().isDocker());
+        String password = "password";
+        Path dockerKeystore = installation.config("elasticsearch.keystore");
+
+        Path localKeystoreFile = getKeystoreFileFromDockerContainer(password, dockerKeystore);
+
+        // restart ES with password and mounted keystore
+        Map<Path, Path> volumes = Map.of(localKeystoreFile, dockerKeystore);
+        Map<String, String> envVars = Map.of("KEYSTORE_PASSWORD", password);
+        runContainer(distribution(), volumes, envVars);
+    }
+
+    /**
+     * Check that if we provide the wrong password for a mounted and password-protected
+     * keystore, Elasticsearch doesn't start.
+     */
+    public void test61DockerEnvironmentVariableBadPassword() throws Exception {
+        assumeTrue(distribution().isDocker());
+        String password = "password";
+        Path dockerKeystore = installation.config("elasticsearch.keystore");
+
+        Path localKeystoreFile = getKeystoreFileFromDockerContainer(password, dockerKeystore);
+
+        // restart ES with password and mounted keystore
+        Map<Path, Path> volumes = Map.of(localKeystoreFile, dockerKeystore);
+        Map<String, String> envVars = Map.of("KEYSTORE_PASSWORD", "wrong");
+        Shell.Result r = runContainerExpectingFailure(distribution(), volumes, envVars);
+        assertThat(r.stderr, containsString(PASSWORD_ERROR_MESSAGE));
+    }
+
+    /**
+     * In the Docker context, it's a little bit tricky to get a password-protected
+     * keystore. All of the utilities we'd want to use are on the Docker image.
+     * This method mounts a temporary directory to a Docker container, password-protects
+     * the keystore, and then returns the path of the file that appears in the
+     * mounted directory (now accessible from the local filesystem).
+     */
+    private Path getKeystoreFileFromDockerContainer(String password, Path dockerKeystore) throws IOException {
```

Comment:
Future thought: should we have a test utility to grab any file from the container, not just the keystore?

Reply:
We already do — `Docker#copyFromContainer`. But because I was already using a mounted directory to load the script, it seemed like a fairly clean way to get this specific file.